# Telco Customer Churn Analysis

This notebook focuses on preparing the raw customer churn dataset
for analysis and modeling. The goal is to create a clean, consistent,
and reproducible dataset for subsequent steps.

The project Github repository is [kushnertodd/telco-customer-churn](https://github.com/kushnertodd/telco-customer-churn).

This notebook is based on the following Towards Data Science project:     
[End-to-end machine learning project: Telco customer churn](https://towardsdatascience.com/end-to-end-machine-learning-project-telco-customer-churn-90744a8df97d/)

The dataset is derived from the [IBM Cognos Analytics 11.1.3+ Telco customer churn samples data module](https://community.ibm.com/community/user/blogs/steven-macko/2019/07/11/telco-customer-churn-1113). 


## Data dictionary
These are the full set of data field descriptions for the IBM Cognos Analytics 11.1.3+ Telco customer churn samples data module. The samples data is located in the IBM Cognos Analytics installation under Team Content > Samples > Data in 5 Excel spreadsheets:

- Telco_customer_churn_demographics.xlsx
- Telco_customer_churn_location.xlsx
- Telco_customer_churn_population.xlsx
- Telco_customer_churn_services.xlsx
- Telco_customer_churn_status.xlsx

Each table field is described below.
<h4>Demographics</h4>
<table style="width:100%">
<tr><th>Field</th><th>Description</th><th>Values</th></tr>
<tr><td>CustomerID</td><td>A unique ID that identifies each customer.</td><td></td></td></tr>
<tr><td>Gender</td><td>The customer's gender.</td><td>Male, Female</td></tr>
<tr><td>Age</td><td>The customer's current age, in years, at the time the fiscal quarter ended.</td><td></td></tr>
<tr><td>Under 30</td><td>Indicates if the customer is under 30 years old.</td><td>Yes, No</td></tr>
<tr><td>Senior Citizen</td><td>Indicates if the customer is 65 or older.</td><td>Yes, No</td></tr>
<tr><td>Married</td><td>Indicates if the customer is married.</td><td>Yes, No</td></tr>
<tr><td>Dependents</td><td>Indicates if the customer lives with any dependents. Dependents could be children, parents, grandparents, etc.</td><td>Yes, No</td></tr>
<tr><td>Number of Dependents</td><td>Indicates the number of dependents that live with the customer.</td><td></td></tr>
</table>
<h4>Location</h4>
<table style="width:100%">
<tr><th>Field</th><th>Description</th><th>Values</th></tr>
<tr><td>Country</td><td>The country of the customer's primary residence.</td><td></td></tr>
<tr><td>State</td><td>The state of the customer's primary residence.</td><td></td></tr>
<tr><td>City</td><td>The city of the customer's primary residence.</td><td></td></tr>
<tr><td>Zip Code</td><td>The zip code of the customer's primary residence.</td><td></td></tr>
<tr><td>Lat Long</td><td>The combined latitude and longitude of the customer's primary residence.</td><td></td></tr>
<tr><td>Latitude</td><td>The latitude of the customer's primary residence.</td><td></td></tr>
<tr><td>Longitude</td><td>The longitude of the customer's primary residence.</td><td></td></tr>
</table>
<h4>Population</h4>
<table style="width:100%">
<tr><th>Field</th><th>Description</th><th>Values</th></tr>
<tr><td>Population</td><td>Estimated population in the customer’s Zip Code area.</td><td></td></tr>
</table>
<h4>Services</h4>
<table style="width:100%">
<tr><th>Field</th><th>Description</th><th>Values</th></tr>
<tr><td>Quarter</td><td>The fiscal quarter that the data has been derived from (e.g. Q3).</td><td></td></tr>
<tr><td>Referred a Friend</td><td>Indicates if the customer's ever referred a friend or family member to this company.</td><td>Yes, No</td></tr>
<tr><td>Number of Referrals</td><td>Indicates the number of referrals to date that the customer's made.</td><td></td></tr>
<tr><td>Tenure in Months</td><td>Indicates the total amount of months that the customer's been with the company by the end of the quarter specified above.</td><td></td></tr>
<tr><td>Offer</td><td>Identifies the last marketing offer that the customer accepted, if applicable.</td><td>None, Offer A, Offer B, Offer C, Offer D, and Offer E</td></tr>
<tr><td>Phone Service</td><td>Indicates if the customer subscribes to home phone service with the company.</td><td>Yes, No</td></tr>
<tr><td>Avg Monthly Long Distance Charges</td><td>Indicates the customer's average long distance charges, calculated to the end of the quarter specified above.</td><td></td></tr>
<tr><td>Multiple Lines</td><td>Indicates if the customer subscribes to multiple telephone lines with the company.</td><td>Yes, No</td></tr>
<tr><td>Internet Service</td><td>Indicates if the customer subscribes to Internet service with the company.</td><td>Yes, No</td></tr>
<tr><td>Internet Type</td><td>Internet service type</td><td>No, DSL, Fiber Optic, Cable</td></tr>
<tr><td>Avg Monthly GB Download</td><td>Indicates the customer's average download volume in gigabytes, calculated to the end of the quarter specified above.</td><td></td></tr>
<tr><td>Online Security</td><td>Indicates if the customer subscribes to an additional online security service provided by the company.</td><td>Yes, No</td></tr>
<tr><td>Online Backup</td><td>Indicates if the customer subscribes to an additional online backup service provided by the company.</td><td>Yes, No</td></tr>
<tr><td>Device Protection Plan</td><td>Indicates if the customer subscribes to an additional device protection plan for their Internet equipment provided by the company.</td><td>Yes, No</td></tr>
<tr><td>Premium Tech Support</td><td>Indicates if the customer subscribes to an additional technical support plan from the company with reduced wait times.</td><td>Yes, No</td></tr>
<tr><td>Streaming TV</td><td>Indicates if the customer uses their Internet service to stream television programing from a third party provider. The company does not charge an additional fee for this service.</td><td>Yes, No</td></tr>
<tr><td>Streaming Movies</td><td>Indicates if the customer uses their Internet service to stream movies from a third party provider. The company does not charge an additional fee for this service.</td><td>Yes, No</td></tr>
<tr><td>Streaming Music</td><td>Indicates if the customer uses their Internet service to stream music from a third party provider. The company does not charge an additional fee for this service.</td><td>Yes, No</td></tr>
<tr><td>Unlimited Data</td><td>Indicates if the customer's paid an additional monthly fee to have unlimited data downloads/uploads.</td><td>Yes, No</td></tr>
<tr><td>Contract</td><td>Indicates the customer's current contract type.</td><td>Month-to-Month, One Year, Two Year</td></tr>
<tr><td>Paperless Billing</td><td>Indicates if the customer's chosen paperless billing.</td><td>Yes, No</td></tr>
<tr><td>Payment Method</td><td>Indicates how the customer pays their bill.</td><td>Bank Withdrawal, Credit Card, Mailed Check</td><td></td></tr>
<tr><td>Monthly Charge</td><td>Indicates the customer's current total monthly charge for all their services from the company.</td><td></td></tr>
<tr><td>Total Charges</td><td>Indicates the customer's total charges, calculated to the end of the quarter specified above.</td><td></td></tr>
<tr><td>Total Refunds</td><td>Indicates the customer's total refunds, calculated to the end of the quarter specified above.</td><td></td></tr>
<tr><td>Total Extra Data Charges</td><td>Indicates the customer's total charges for extra data downloads above those specified in their plan, by the end of the quarter specified above.</td><td></td></tr>
<tr><td>Total Long Distance Charges</td><td>Indicates the customer's total charges for long distance above those specified in their plan, by the end of the quarter specified above.</td><td></td></tr>
<tr><td>Total Revenue</td><td>Indicates the total customer charges.</td><td></td></tr>
</table>
<h4>Status</h4>
<table style="width:100%">
<tr><th>Field</th><th>Description</th><th>Values</th></tr>
<tr><td>Satisfaction Score</td><td>A customer's overall satisfaction rating of the company.</td><td>from 1 (Very Unsatisfied) to 5 (Very Satisfied)</td></tr>
<tr><td>Customer Status</td><td>Indicates the status of the customer at the end of the quarter.</td><td>Churned, Stayed, or Joined</td></tr>
<tr><td>Churn Label</td><td>Indicates whether the customer churned this quarter.</td><td>Yes = the customer left the company this quarter<br>No = the customer remained with the company</td></tr>
<tr><td>Churn Score</td><td>A value from 0-100 that is calculated using the predictive tool IBM SPSS Modeler. The model incorporates multiple factors known to cause churn. The higher the score, the more likely the customer will churn.</td><td></td></tr>
<tr><td>CLTV</td><td>Customer Lifetime Value. A predicted CLTV is calculated using corporate formulas and existing data. The higher the value, the more valuable the customer. High value customers should be monitored for churn.</td><td></td></tr>
<tr><td>Churn Category</td><td>A high-level category for the customer's reason for churning. When they leave the company, all customers are asked about their reasons for leaving. Directly related to Churn Reason.</td><td>Attitude, Competitor, Dissatisfaction, Other, Price</td><td></td></tr>
<tr><td>Churn Reason</td><td>A customer's specific reason for leaving the company. Directly related to Churn Category.</td><td></td></tr>
</table>


## Telco Customer Churn Analysis dataset

A full dataset can be found in `telco.csv`. A version reduced to the useful model variables in [WA_Fn-UseC_-Telco-Customer-Churn.csv](https://www.kaggle.com/datasets/blastchar/telco-customer-churn) is used: 
- Customer ID
- Gender
- Senior Citizen
- Married
- Dependents
- Tenure in Months
- Phone Service
- Multiple Lines
- Internet Type
- Online Security
- Online Backup
- Device Protection Plan
- Premium Tech Support
- Streaming TV
- Streaming Movies
- Contract
- Paperless Billing
- Payment Method
- Monthly Charge
- Total Charges
- Customer Status


## Importing Required Libraries

Pandas is used for data manipulation and analysis.
Display formatting is adjusted to improve readability
when inspecting dataframes.


In [1]:
import pandas as pd
# Display formatting
from IPython.core.display import HTML
HTML("""
<style>
.dataframe table, .dataframe th, .dataframe td { font-size: 12px; }
div.output_scroll { overflow-x: auto; }
</style>
""")

### Loading the Dataset

The dataset is loaded from the raw data directory.
This raw version is kept unchanged to ensure data integrity and reproducibility.


In [2]:
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

## Initial Data Inspection


In [3]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


Shape of the dataset

In [4]:
df.shape

(7043, 21)

Check data types

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


Checking the number of unique values of each column

In [6]:
df.nunique()

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

Check statistics of data set

In [7]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


### Initial Data Inspection

At this stage, we inspect:
- The number of rows and columns
- Sample records
- Column names and data types
- Presence of missing values

This step helps build an initial understanding of the dataset before making any transformations.


### Standardizing Column Names

Column names are standardized to improve code readability and consistency.
The following transformations are applied:
- Convert all names to lowercase
- Replace spaces with underscores
- Remove special characters

This step helps prevent errors during analysis and modeling.


In [9]:
df.columns = (
    df.columns
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('-', '_')
)

df.columns

Index(['customerid', 'gender', 'seniorcitizen', 'partner', 'dependents',
       'tenure', 'phoneservice', 'multiplelines', 'internetservice',
       'onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport',
       'streamingtv', 'streamingmovies', 'contract', 'paperlessbilling',
       'paymentmethod', 'monthlycharges', 'totalcharges', 'churn'],
      dtype='object')

### Reviewing Data Types

The dataset contains a mix of numerical and categorical variables.
Many categorical features are currently stored as object types.

At this stage, no type conversion is performed.
Detailed encoding and transformations will be handled during the feature engineering phase.


In [10]:
df.dtypes.value_counts()

object     18
int64       2
float64     1
Name: count, dtype: int64

### Identifying Missing Values

Missing values are inspected across all columns.
Some missing values represent meaningful business information rather than data errors.
Therefore, missing values will be handled carefully instead of removing rows.

In [11]:
df.isna().sum().sort_values(ascending=False)

customerid          0
deviceprotection    0
totalcharges        0
monthlycharges      0
paymentmethod       0
paperlessbilling    0
contract            0
streamingmovies     0
streamingtv         0
techsupport         0
onlinebackup        0
gender              0
onlinesecurity      0
internetservice     0
multiplelines       0
phoneservice        0
tenure              0
dependents          0
partner             0
seniorcitizen       0
churn               0
dtype: int64

### Handling Missing Categorical Values

Missing values in categorical columns such as `offer` and `internet_type` are replaced with explicit labels.
This approach preserves all records and allows the model to learn from the absence of these services or offers.
This approach avoids unnecessary data loss
and preserves meaningful customer information.


In [19]:
df['offer'] = df['offer'].fillna('No Offer')
df['internet_type'] = df['internet_type'].fillna('No Internet')

### Insights from Numerical Data
After examining the descriptive statistics of the numerical features, the following insights can be drawn:

Age:

- Mean age of customers is ~46.5 years.
- Standard deviation is ~16.75, indicating a moderately wide spread.
- Youngest customer is 19 and oldest is 80, showing a diverse age range.

Number of Dependents:
- Most customers have 0 dependents (median = 0).
- Maximum number of dependents is 9, but mean is only ~0.47, suggesting most customers live alone or with few dependents.

Location (Zip Code, Latitude, Longitude, Population):
- Zip Codes range from 90001 to 96150.
- Latitude and Longitude indicate coverage across multiple states.
- Population varies widely from 11 to 105,285 per area.

Customer Engagement:
- Number of Referrals: Most customers have referred 0 people; maximum referrals = 11.
- Tenure in Months: Average tenure is ~32 months, ranging from 1 to 72 months.

Service Usage and Billing:
- Avg Monthly Long Distance Charges: Mean ~22.96, maximum ~50.
- Avg Monthly GB Download: Mean ~20.52 GB; some customers download up to 85 GB.
- Monthly Charges: Average monthly bill ~64.76, with a max of 118.75.
- Total Charges: Mean ~2,280, but max ~8,684, showing some high-value customers.
- Total Refunds, Extra Data, Long Distance Charges: Some customers have unusually high extra charges, suggesting outliers.

Customer Metrics:
- Satisfaction Score: Average score ~3.24, ranging 1–5.
- Churn Score: Mean ~58.5, max 96, showing that some customers are highly likely to churn.
- CLTV (Customer Lifetime Value): Average ~4,400; high CLTV customers (max 6,500) are critical for retention.

Observations:

- Many numerical features are moderately skewed (e.g., Tenure, Total Charges, Extra Charges).
- There are clear outliers in several columns (Population, Total Charges, Extra Data Charges, Long Distance Charges).
- Most customers have low numbers of dependents and referrals.
- High churn risk customers can be identified using Churn Score and high-value customers using CLTV.

In [37]:
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 19 numerical features : ['age', 'number_of_dependents', 'zip_code', 'latitude', 'longitude', 'population', 'number_of_referrals', 'tenure_in_months', 'avg_monthly_long_distance_charges', 'avg_monthly_gb_download', 'monthly_charge', 'total_charges', 'total_refunds', 'total_extra_data_charges', 'total_long_distance_charges', 'total_revenue', 'satisfaction_score', 'churn_score', 'cltv']

We have 31 categorical features : ['customer_id', 'gender', 'under_30', 'senior_citizen', 'married', 'dependents', 'country', 'state', 'city', 'quarter', 'referred_a_friend', 'offer', 'phone_service', 'multiple_lines', 'internet_service', 'internet_type', 'online_security', 'online_backup', 'device_protection_plan', 'premium_tech_support', 'streaming_tv', 'streaming_movies', 'streaming_music', 'unlimited_data', 'contract', 'paperless_billing', 'payment_method', 'customer_status', 'churn_label', 'churn_category', 'churn_reason']


### Saving the Cleaned Dataset

The cleaned dataset is saved to the processed data directory.
This version will be used for exploratory data analysis and modeling.


In [20]:
df.to_csv('cleaned_data.csv', index=False)